haskell/test/Test/Ligo/BaseDAO/Proposal/Propose.hs

```diff
@@ -532,3 +536,66 @@ proposalBoundedValue originateFn = do
     call dodDao (Call @"Propose") params
       & expectFailedWith maxProposalsReached
 
+
+unstakeVote
+  :: (MonadCleveland caps base m, HasCallStack)
+  => (ConfigDesc Config -> OriginateFn m)
+  -> m ()
+unstakeVote originateFn = do
+  DaoOriginateData{..}
+    <- originateFn (testConfig
+        >>- (ConfigDesc configConsts{ cmProposalFlushTime = Just 20 })
+        >>- (ConfigDesc configConsts{ cmProposalExpiredTime = Just 30 })
+        ) defaultQuorumThreshold
+  originationLevel <- getOriginationLevel dodDao
+
+  -- [Voting]
+  withSender dodOwner1 $
+    call dodDao (Call @"Freeze") (#amount :! 30)
+
+  withSender dodOwner2 $
+    call dodDao (Call @"Freeze") (#amount :! 10)
+
+  startLevel <- getOriginationLevel dodDao
+  advanceToLevel (startLevel + dodPeriod)
+
+  -- [Proposing]
+  key1 <- createSampleProposal 1 dodOwner1 dodDao
+  advanceLevel 1
+  key2 <- createSampleProposal 2 dodOwner1 dodDao
+
+  let vote' key = NoPermit VoteParam
+        { vFrom = dodOwner2
+        , vVoteType = True
+        , vVoteAmount = 5
+        , vProposalKey = key
+        }
+
+  advanceToLevel (originationLevel + 2 * dodPeriod)
+
+  -- [Voting]
+  withSender dodOwner2 . inBatch $ do
+      call dodDao (Call @"Vote") [vote' key1]
+      call dodDao (Call @"Vote") [vote' key2]
+      pure ()
+
+  proposalStart2 <- getProposalStartLevel dodDao key2
+  advanceToLevel (proposalStart2 + 21)
+
+  -- [Proposing]
+  withSender dodAdmin $ call dodDao (Call @"Flush") 1
+
+  -- Call succeeds.
+  withSender dodOwner2 $ call dodDao (Call @"Unstake_vote") [key1]
+
+  -- Trigger error since the tokens are already unstaked.
+  withSender dodOwner2 $ call dodDao (Call @"Unstake_vote") [key1]
+    & expectFailedWith voterAlreadyUnstake
+
+  -- Proposal key2 is not yet flush.
+  withSender dodOwner2 $ call dodDao (Call @"Unstake_vote") [key2]
+    & expectFailedWith proposalIsNotFlushed
+
+  -- owner 1 does not vote on the proposal
+  withSender dodOwner1 $ call dodDao (Call @"Unstake_vote") [key1]
+    & expectFailedWith voterDoesNotExist
```

Comment:
AFAIU this test does not actually test if the "Unstaking" worked. Right?

Reply:
I added `freeze_history` comparison now, this should test if unstaking works properly.

Reply:
Can we also add a test to check that the unstaking works when the proposal is dropped?

Reply:
Make sense, I've added a case for when the proposal is dropped as well.